csrc/index_compute.cpp

```diff
@@ -3291,81 +3291,571 @@ Val* Index::eye(
+    BestEffortReplay replay_producer_as_consumer(
+        producer_tv->getLeafDomain(),
+        consumer_tv->getLeafDomain(),
+        c2p_root_map);
+
+    const auto& c2p_map = replay_producer_as_consumer.getReplay();
+
+    consumer_to_gmem = [=](IterDomain* id) -> IterDomain* {
+      return c2p_map.at(id);
+    };
+
+    const auto& producer_indexing_from_idgraph = getTensorIndexFromIdGraph(
+        loops, rotated_loops, consumer_tv, producer_tv, true, c2p_map);
+
+    indexing =
+        std::make_unique<IndexCompute>(producer_indexing_from_idgraph.index);
+  } else {
+    consumer_to_gmem = [](IterDomain* id) -> IterDomain* { return id; };
+    auto index_from_id_graph =
+        getTensorIndexFromIdGraph(loops, rotated_loops, consumer_tv);
+    indexing = std::make_unique<IndexCompute>(index_from_id_graph.index);
+  }
+
+  // Step 1: Get all bulk IterDomains and originating bulk IterDomains
+
+  // Get all bulk IterDomains
+  std::unordered_set<IterDomain*> bulk_ids;
+  // Bulk IterDomains that we need to check its definition to see if it is an
+  // originating bulk IterDomain.
+  std::deque<IterDomain*> pending;
+  pending.push_back(nullptr); // use nullptr as a checkpoint
+  // Start from leaf domain, where all the bulk IterDomains in the leaf domain
+  // must be parallelized as ParallelType::Bulk.
+  for (auto id : consumer_tv->getLeafDomain()) {
+    if (id->getParallelType() == ParallelType::Bulk) {
+      id = consumer_to_gmem(id);
+      bulk_ids.insert(id);
+      pending.push_back(id);
+    }
+  }
+  // Use a BFS-like (not exactly BFS) algorithm to propagate back to get all
+  // bulk IterDomains
+  bool updated = true;
+  while (true) {
+    auto id = pending.front();
+    pending.pop_front();
+    if (id == nullptr) {
+      if (updated) {
+        // We discovered new bulk IterDomains in the last round, so we need to
+        // continue start a new round to see if we can discover more bulk
+        // IterDomains.
+        pending.push_back(nullptr);
+        updated = false;
+        continue;
+      } else {
+        // We have visited all IterDomains in pending for one round, but nothing
+        // has changed. This means that all IterDomains in pending are
+        // originating bulk IterDomains, so we can no longer propagate further.
+        break;
+      }
+    }
+    auto def = id->definition();
+    NVF_ERROR(
+        def != nullptr,
+        "A bulk IterDomain must have a definition, but ",
+        id,
+        " does not have one.");
+    if (bulk_ids.count(def->input(0)->as<IterDomain>()) > 0) {
+      // already processed from another path
+      continue;
+    }
+
+    bool all_outputs_are_bulk = true;
+    for (auto out : def->outputs()) {
+      if (bulk_ids.count(out->as<IterDomain>()) == 0) {
+        all_outputs_are_bulk = false;
+        break;
+      }
+    }
 
+    if (all_outputs_are_bulk) {
+      // All outputs of def are bulk IterDomains, so all inputs of def are bulk
```

Comment:
> inputs will be bulk due to the fact that we are replaying transformations from consumer to producer? 

No, this has nothing to do with replaying producer as consumer. We replay producer as consumer because the consumer dictate the loop structure of an expression. For example, if we have `T1[I1*I2] = T0[I1, I2]`, we will have loop structure

```python
for i in I1*I2:
  T1[i] = T0[i / size2, i % size2]
```

instead of

```python
for i1 in I1:
  for i2 in I2:
    T1[i1 * size2 + i2] = T0[i1, i2]
```

Reply:
I think the confusion here is the concept "bulk IterDomain" vs "IterDomain parallelized as `ParallelType::Bulk`", they are different concepts. An "IterDomain parallelized as `ParallelType::Bulk`" is always a "bulk IterDomain", but not the reverse. We only parallelize leaf domains, so in your first diagram, it is an error if you parallelize `I2` as `ParallelType::Bulk`. It is a "bulk IterDomain" but not an "IterDomain parallelized as `ParallelType::Bulk`".

Reply:
I added some texts explaining these two concepts: https://github.com/NVIDIA/Fuser/pull/1962/commits/bfa3ddbcca8d24587b38b3427d0a1076c35d256a

Reply:
Thank you @zasdfgbnm for explanation, especially for adding explanation about the difference between "bulk IterDomain" and "IterDomain parallelized as Bulk".

In the first diagram, I though this operation is legal, and the aim of such operation was to set parallelization type once, and have it automatically propagate to ancestor IterDomains, created when we apply merge/split.